sympy/polys/matrices/rref.py

```diff
@@ -0,0 +1,551 @@
+        :ref:`ZZ` or :ref:`QQ`.
+
+        With ``method='GJ'`` the matrix is converted to the associated field
+        domain and Gauss-Jordan elimination with division is used. With
+        ``method='FF'`` fraction-free Gauss-Jordan elimination is used in the
+        current domain and the result is converted to the associated field
+        domain at the end. With ``method='CD'`` the denominators are cleared
+        and fraction-free Gauss-Jordan elimination is used in the associated
+        ring before converting to the associated field domain and dividing at
+        the end.
+
+        By default the sparse implementations of the different algorithms are
+        used. To use the dense implementations instead use
+        ``method='GJ_dense'``, ``method='FF_dense'``, or ``method='CD_dense'``.
+        With ``method='auto'`` only the sparse implementations are used.
+
+        In all cases the result is returned over the field associated with the
+        domain of the matrix and the format of the returned matrix (sparse or
+        dense) is always the same as the format of the input matrix.
+
+    Returns
+    =======
+
+    M_rref : DomainMatrix
+        The reduced row echelon form of the matrix over the field associated
+        with the domain of the matrix.
+    pivots : tuple
+        The indices of the pivot columns.
+
+    See Also
+    ========
+
+    sympy.polys.matrices.domainmatrix.DomainMatrix.rref
+        The ``DomainMatrix`` method that calls this function.
+    sympy.polys.matrices.rref.dm_rref_den
+        Alternative function for computing RREF with denominator.
+    """
+    K = M.domain
+
+    is_dense = M.rep.fmt == 'dense'
+
+    # Do not switch to the sparse implementation for EX because the domain does
+    # not have proper canonicalization and the sparse implementation gives
+    # equivalent but non-identical results over EX from performing arithmetic
+    # in a different order. Specifically test_issue_23718 ends up getting a
+    # more complicated expression when using the sparse implementation.
+    # Probably the best fix for this is something else but for now we stick
+    # with the dense implementation for EX if the matrix is already dense.
+    if method == 'auto' and K.is_EX and is_dense:
+        return _dm_rref_GJ_dense(M)
+
+    if method == 'auto':
+        method, use_dense = _dm_rref_choose_method(M)
+    else:
+        use_dense = method.endswith('_dense')
+        if use_dense:
+            method = method[:-len('_dense')]
+
+    if use_dense:
+        if not is_dense:
+            M = M.to_dense()
+
+        M_rref, pivots = _dm_rref(M, method, use_dense)
+
+        if not is_dense:
+            M_rref = M_rref.to_sparse()
+
+    else:
+        if is_dense:
+            M = M.to_sparse()
+
+        M_rref, pivots = _dm_rref(M, method, use_dense)
+
+        if is_dense:
+            M_rref = M_rref.to_dense()
+
+    return M_rref, pivots
```

Comment:
I wonder if things are less messy to copy everything, if there is no `_den` stuff and `den` can be just `one`